File: src/analyzer/reference-call-receiver.ts
import type { AstNode } from "../ir/primitive/ast-node.js";
import { AST_TYPE } from "../parser/ast-type.js";

export function referenceCallReceiverFlags(
  parent: AstNode | null,
  key: string | null,
): Readonly<{ call: boolean; receiver: boolean }> {
  if (parent === null) {
    return { call: false, receiver: false };
  }
  const t = parent.type;
  if (t === AST_TYPE.CallExpression && key === "callee") {
    return { call: true, receiver: false };
  }
  if (t === AST_TYPE.NewExpression && key === "callee") {
    return { call: true, receiver: false };
  }
  if (t === AST_TYPE.MemberExpression && key === "object") {
    return { call: false, receiver: true };
  }
  return { call: false, receiver: false };
}

File: crates/unsnarl-analyzer/src/reference_call_receiver.rs
//! Compute `call` / `receiver` flags for a reference.
//!
//! Mirrors `ts/src/analyzer/reference-call-receiver.ts`. The flags
//! report whether a reference appears as the callee of a call /
//! `new`, or as the object of a member expression.

use unsnarl_annotations::ReferenceAnnotationFlags;
use unsnarl_oxc_parity::AstType;

pub fn reference_call_receiver_flags(
    parent_type: Option<&AstType>,
    key: Option<&str>,
) -> ReferenceAnnotationFlags {
    let Some(parent_type) = parent_type else {
        return ReferenceAnnotationFlags {
            call: false,
            receiver: false,
        };
    };
    match (parent_type, key) {
        (AstType::CallExpression, Some("callee")) | (AstType::NewExpression, Some("callee")) => {
            ReferenceAnnotationFlags {
                call: true,
                receiver: false,
            }
        }
        (AstType::MemberExpression, Some("object")) => ReferenceAnnotationFlags {
            call: false,
            receiver: true,
        },
        _ => ReferenceAnnotationFlags {
            call: false,
            receiver: false,
        },
    }
}

#[cfg(test)]
#[path = "reference_call_receiver_test.rs"]
mod reference_call_receiver_test;
